Date: Fri, 9 Jun 2023 11:49:58 -0700
Subject: crypto: Only allow GCM in FIPS when instantiated via seqiv

FIPS 140-3 requires that GCM can be approved IFF it is instantiated via
seqiv. Set a new flag CRYPTO_ALG_FIPS140_COMPLIANT in a separate init
function aead_init_seqiv() to represent that seqiv was used instead of
geniv.

Signed-off-by: Samuel Mendoza-Jonas <samjonas@amazon.com>
Signed-off-by: Elena Avila <ellavila@amazon.com>

diff --git a/crypto/seqiv.c b/crypto/seqiv.c
--- a/crypto/seqiv.c
+++ b/crypto/seqiv.c
@@ -134,6 +134,12 @@ static int seqiv_aead_decrypt(struct aead_request *req)
 	return crypto_aead_decrypt(subreq);
 }
 
+static int aead_init_seqiv(struct crypto_aead *aead)
+{
+	crypto_aead_set_flags(aead, CRYPTO_ALG_FIPS140_COMPLIANT);
+	return aead_init_geniv(aead);
+}
+
 static int seqiv_aead_create(struct crypto_template *tmpl, struct rtattr **tb)
 {
 	struct aead_instance *inst;
@@ -151,7 +157,7 @@ static int seqiv_aead_create(struct crypto_template *tmpl, struct rtattr **tb)
 	inst->alg.encrypt = seqiv_aead_encrypt;
 	inst->alg.decrypt = seqiv_aead_decrypt;
 
-	inst->alg.init = aead_init_geniv;
+	inst->alg.init = aead_init_seqiv;
 	inst->alg.exit = aead_exit_geniv;
 
 	inst->alg.base.cra_ctxsize = sizeof(struct aead_geniv_ctx);
diff --git a/include/linux/crypto.h b/include/linux/crypto.h
--- a/include/linux/crypto.h
+++ b/include/linux/crypto.h
@@ -124,6 +124,12 @@
  */
 #define CRYPTO_ALG_FIPS_INTERNAL	0x00020000
 
+/*
+ * Mark an algorithm as approved for FIPS 140-3. This is intended to be used
+ * for algorithms that are only approved under certain conditions
+ */
+#define CRYPTO_ALG_FIPS140_COMPLIANT	0x00040000
+
 /*
  * Transform masks and values (for crt_flags).
  */
